### PR TITLE
COMPAT: Expand compatibility fromnumeric.py

### PR DESCRIPTION
**Pull request by gfyoung**
Addresses issues in #5987 by fixing signature incompatibilities between functions defined in `numpy.core.fromnumeric` and similarly defined ones for `scipy.sparse` matrices by expanding the signatures in `scipy.sparse` to accommodate (and also implement) some of the functionality offered in `numpy`, in particular, these four functions:

1) `sum`
2) `mean`
3) `reshape`
4) `transpose`

cc @perimosocordiae

### CODE REVIEW

**Comment by gfyoung** [comment]
1) Why isn't `reshape` implemented for the other sparse matrices?
2) `np.diagonal` is another one in which there is a signature incompatibility, but it doesn't surface because `numpy` casts everything to `ndarray` (which is disastrous for a sparse matrix but no other `Exception` is thrown at the very least).  Is it worthwhile to match signatures even though it technically not an issue (aside from the casting part)?

File: scipy/sparse/__init__.py [hunk 1]
```diff
@@ -103,9 +103,16 @@
     7. dia_matrix: DIAgonal format
 
 To construct a matrix efficiently, use either dok_matrix or lil_matrix.
-The lil_matrix class supports basic slicing and fancy
-indexing with a similar syntax to NumPy arrays.  As illustrated below,
-the COO format may also be used to efficiently construct matrices.
+The lil_matrix class supports basic slicing and fancy indexing with a
+similar syntax to NumPy arrays. As illustrated below, the COO format
+may also be used to efficiently construct matrices. Despite their
+to NumPy arrays, it is **strongly discouraged** to use NumPy functions
```
Comment (perimosocordiae) [inline scipy/sparse/__init__.py]: missing a word here?

**Comment by perimosocordiae** [comment]
I'm not sure if pointing to the numpy functions in the "See Also" block is worthwhile. I would rather point to the corresponding ndarray method, or just omit that section entirely.

There have been efforts to implement a `reshape` method for other sparse formats (see #3957) but motion has stalled. In theory, it should be implemented more broadly.

If `np.diagonal` doesn't attempt to call `spmatrix.diagonal`, then I'd prefer to leave it as-is.

**Comment by gfyoung** [comment]
Okay, very well.  I'll leave `spmatrix.diagonal` as is then.  I think actually it would better on second thought to point to the `np.matrix` methods since that is what we return for the methods that I mentioned.

With regards to `reshape`, yikes!  In my mind, it felt like it should be quite simple given the abundance of conversion functions we have lying around that we use quite freely for `transpose` for example.  I am perfectly okay with not implementing them for the other methods given that that was not the original purpose of my PR.  Some feedback on my implementation decision for `order` would be nice, since that could have consequences on future implementations of the method for other sparse matrix types.

File: scipy/sparse/__init__.py (hunk 1 above)
Comment (gfyoung) [inline scipy/sparse/__init__.py]: Good catch.  Fixed.

File: scipy/sparse/base.py [hunk 2]
```diff
@@ -767,48 +845,129 @@ def sum(self, axis=None):
...
-            return (self * np.asmatrix(np.ones((n, 1), dtype=res_dtype))).sum()
+            return (self * np.asmatrix(np.ones(
+                (n, 1), dtype=res_dtype))).sum(
+                dtype=dtype, out=out)
+
+        axis_type = type(axis)
+
+        # In NumPy, you can pass in tuples for 'axis', so let's make
+        # it explicit that you cannot do that with sparse matrices
+        if axis_type == tuple:
```
Comment (perimosocordiae) [inline scipy/sparse/base.py]: It might be better to just add support for axis tuples, and avoid the inconsistency. The implementation shouldn't be too complex. We could pull out the code that processes the axis argument into a helper function, because this might be useful in other places as well.

File: scipy/sparse/base.py [hunk 3]
```diff
@@ -537,8 +557,32 @@ def __getattr__(self, attr):
...
+    def transpose(self, axes=None, copy=False):
+        """
+        Reverses the dimensions of the sparse matrix.
+
+        Parameters
+        ----------
+        axes : None, optional
+            This argument is in the signature *solely* for NumPy
+            compatibility reasons, as reversing the dimensions is
+            the only logical permutation that should be requested.
```
Comment (perimosocordiae) [inline scipy/sparse/base.py]: As with `sum`'s `axis` parameter, it might be easier to just implement this. It's not particularly useful to have anything but `(1,0)`, but it matches the `np.matrix` behavior better.

File: scipy/sparse/tests/test_csc.py [hunk 4]
```diff
@@ -35,5 +36,26 @@ def test_csc_getcol():
...
+    A = np.array([[1, 0, 1], [0, 1, 1], [0, 0, 1]])
+    Asp = csc_matrix(A)
+
+    A_nz = set([tuple(ij) for ij in np.transpose(np.nonzero(A))])
+    Asp_nz = set([tuple(ij) for ij in np.transpose(np.nonzero(Asp))])
+
+    assert_equal(A_nz, Asp_nz)
+
+
+if __name__ == "q__main__":
```
Comment (perimosocordiae) [inline scipy/sparse/tests/test_csc.py]: This looks like a typo, but I think the changes to this file aren't needed. The tests in test_base.py should be sufficient.

File: scipy/sparse/base.py (hunk 3 above)
Comment (gfyoung) [inline scipy/sparse/base.py]: The issue that I have is that is the additional flexibility worth the extra validation complexity?  There are AFAICT _thirteen_ "valid" tuples that you can pass in based on the `np.matrix` functionality, and if we're going to support tuples, I would think that all of them need to be accepted.  It seems simpler IMHO to not support them since none of them are actually any more meaningful than the parameters I am currently accepting.

File: scipy/sparse/tests/test_csc.py (hunk 4 above)
Comment (gfyoung) [inline scipy/sparse/tests/test_csc.py]: Good point, and that was indeed a typo.  Fixed.

**Comment by gfyoung** [comment]
xref <a href="https://github.com/numpy/numpy/issues/7460>#7460</a> in `numpy`.  It looks like some more incompatibilities to be squashed, both on the `numpy` and `scipy` ends.

**Comment by perimosocordiae** [comment]
@gfyoung: It's pretty hard to follow this PR because it includes a lot of PEP8/whitespace changes mixed in with the  actual code changes. Could you hold off on squashing for a while to help me follow the progress better?

File: scipy/sparse/base.py (hunk 3 above)
Comment (perimosocordiae) [inline scipy/sparse/base.py]: Eh, I guess it's not a big deal. Out of curiosity, what are the 13 valid tuples? I would expect `(0,1)` and `(1,0)` to work, but that's about it. Unless we're including negative indices here, but those could be dealt with first in the same way that they're handled elsewhere.

**Comment by gfyoung** [comment]
@perimosocordiae : Oh goodness!  I must have done the squashing without noticing!  So sorry!

File: scipy/sparse/base.py (hunk 3 above)
Comment (gfyoung) [inline scipy/sparse/base.py]: Per testing with `np.matrix`:

`()`
`(-2,)`, `(-1,)`, `(0,)`, `(1,)`,
`(-2, -1)`, `(-2, 1)`, `(0, -1)`, `(0, 1)`
`(-1, -2)`, `(-1, 0)`, `(1, -2)`, `(1, 0)`

**Comment by gfyoung** [comment]
To give you a hand at what the most important parts of the PR are, search for any of the following terms:

`def sum`
`def min`
`def max`
`def mean`
`def reshape`
`def transpose`

Again, my bad about throwing all of those PEP8 changes along with it.

**Comment by gfyoung** [comment]
@perimosocordiae : Any updates on this PR?  Again, sorry that I mixed the PEP8 changes with the actual changes, so if that is impeding your ability to review this, I perfectly understand.

File: scipy/sparse/__init__.py (hunk 1 above)
Comment (perimosocordiae) [inline scipy/sparse/__init__.py]: Has this been fixed?

Comment (gfyoung) [inline scipy/sparse/__init__.py]: Ummm...that's really awkward.  I swore I fixed it (otherwise, I wouldn't have made that comment above).  It should be fixed for good now.

File: scipy/sparse/base.py [hunk 5]
```diff
@@ -100,7 +100,26 @@ def get_shape(self):
...
+        """
+        Gives a new shape to a sparse matrix without changing its data.
+
+        Parameters
+        ----------
+        shape : length-2 tuple of ints
+            The new shape should be compatible with the original shape.
+        order : 'C', optional
+            This argument is in the signature *solely* for NumPy
+            compatibility reasons for the time being.
```
Comment (perimosocordiae) [inline scipy/sparse/base.py]: I don't see `order` ever being used. I'm not sure exactly how to signal that clearly in the docs, though. Maybe we can find some existing examples of this sort of thing to copy from?

Comment (gfyoung) [inline scipy/sparse/base.py]: The `pandas` library has dealt with this issue before.  That's where I based the documentation.

**Comment by gfyoung** [comment]
@perimosocordiae : Any updates on this PR?  Again, sorry that I mixed the PEP8 changes with the actual changes, so if that is impeding your ability to review this, I perfectly understand.

**Comment by gfyoung** [comment]
@perimosocordiae : Any updates on this PR?  Again, sorry that I mixed the PEP8 changes with the actual changes, so if that is impeding your ability to review this, I perfectly understand.

**Comment by gfyoung** [comment]
@perimosocordiae : Any updates on this PR?  Again, sorry that I mixed the PEP8 changes with the actual changes, so if that is impeding your ability to review this, I perfectly understand.

**Comment by gfyoung** [comment]
@perimosocordiae : Any updates on this PR?  Again, sorry that I mixed the PEP8 changes with the actual changes, so if that is impeding your ability to review this, I perfectly understand.

**Comment by gfyoung** [comment]
Can somebody take a look at this?  It seems like @perimosocordiae is very busy ATM, but I do see other people with merge-powers actively merging things.  It would be great if someone could take a look at this, as it is a compatibility issue with `numpy`.

File: scipy/sparse/base.py [hunk 6]
```diff
@@ -100,7 +101,26 @@ def get_shape(self):
...
+        """
+        Gives a new shape to a sparse matrix without changing its data.
+
+        Parameters
+        ----------
+        shape : length-2 tuple of ints
+            The new shape should be compatible with the original shape.
+        order : 'C', optional
+            This argument is in the signature *solely* for NumPy
+            compatibility reasons.
```
Comment (Eric89GXL) [inline scipy/sparse/base.py]: I'm not quite sure what this means -- does it mean that it's not used, or that only `'C'` is supported, or...?

Comment (gfyoung) [inline scipy/sparse/base.py]: This means don't pass in anything for `order` and leave it as the default value `'C'` because this argument is only in the signature for compatibility reasons.

Comment (Eric89GXL) [inline scipy/sparse/base.py]: So it's not used, then? The docstring should be updated to clarify

**Comment by Eric89GXL** [comment]
Looks reasonable to me, but I'll let someone more familiar with `scipy.sparse` have a look and merge when they're happy.

File: scipy/sparse/base.py (hunk 6 above)
Comment (gfyoung) [inline scipy/sparse/base.py]: Fixed.

**Comment by gfyoung** [comment]
@Eric89GXL : Do you know of anyone (with merging powers) in particular?  If so, you could perhaps ping them directly in this PR.

**Comment by Eric89GXL** [comment]
I suspect most, if not all, merge-power folks have repo notifications on or look through PRs/issues when they have time. I've tagged this as `scipy.sparse` so that when sorting by submodule it can be found easily.

**Comment by gfyoung** [comment]
@Eric89GXL : Oh, what I meant was do you know anyone in particular who is very familiar with `scipy.sparse`! :smile:

File: scipy/sparse/base.py [hunk 7]
```diff
@@ -537,8 +558,33 @@ def __getattr__(self, attr):
...
+    def transpose(self, axes=None, copy=False):
+        """
+        Reverses the dimensions of the sparse matrix.
+
+        Parameters
+        ----------
+        axes : None, optional
+            This argument is in the signature *solely* for NumPy
+            compatibility reasons. Do not pass in anything except
+            for the default value, as reversing the dimensions is
```
Comment (perimosocordiae) [inline scipy/sparse/base.py]: The phrase starting at the comma on this line is awkward, and can probably be removed entirely.

File: scipy/sparse/data.py [hunk 8]
```diff
@@ -167,31 +172,74 @@ def _min_or_max(self, axis, min_or_max):
...
+        Parameters
+        ----------
+        axis : {-2, -1, 0, 1, None} optional
+            Axis along which the sum is computed. The default is to
+            compute the maximum over all the matrix elements, returning
+            a scalar (i.e. `axis` = `None`).
+
+        out : None, optional
+            This argument is in the signature *solely* for NumPy
+            compatibility reasons for the time being. Do not pass
```
Comment (perimosocordiae) [inline scipy/sparse/data.py]: The phrases "for the time being" and "at the moment" are a bit speculative, and should be removed.

**Comment by perimosocordiae** [comment]
I made a few doc-related comments, but once those are resolved I'm +1 to merge.

Thanks for your patience, @gfyoung!

File: scipy/sparse/base.py (hunk 7 above)
Comment (gfyoung) [inline scipy/sparse/base.py]: Done.

File: scipy/sparse/data.py (hunk 8 above)
Comment (gfyoung) [inline scipy/sparse/data.py]: Done.

**Comment by gfyoung** [comment]
@perimosocordiae @Eric89GXL : Travis has given the green light.  Ready to merge if there is nothing else.

**Comment by perimosocordiae** [comment]
Merged!
